docs/src/references/operation-workflow.md

```diff
@@ -0,0 +1,372 @@
+
+### Operation Implementation
+
+A workflow implementation for a specific operation must implement the actions specified for each non *no-op* state.
+This implementation has some freedom, notably to implement extra checks and actions but also to add new states.
+
+However, there are some rules and best practices.
+
+- All the state messages must be published as retained with QoS 1.
+- A workflow implementation should not reject a state message payload with unknown fields.
+  - It's also important to keep these unknown fields in the following states.
+  - This is important as we want to *extend* the workflow of an operation.
+    A software component added by the user might need these *extra* fields the plugin is not aware of.
+- A workflow implementation must not react on *no-op* states nor terminal states.
+  - The transition from a *no-op* state must be handled either by thin-edge as direct transition
+    or overridden by the user with domain-specific checks and actions.
+  - The terminal states, a.k.a `Successful` and `Failed`, are owned by the process which created the `Init` state (in practice, the mapper).
+    Only this process should clear the retained message state for an operation instance by sending an empty payload on command's topic.
+
+### Workflow Overriding
+
+Thin-edge provides a mechanism to override, extend and combine workflows.
+
+This mechanism is provided by the `tedge-agent` which gather a set of user-defined workflows
+and combined them with the builtin workflows implemented by the agent itself and the thin-edge operation plugins.
+
+Each workflow is defined using a TOML file that specifies:
+- the operation target and name defined as a topic filter
+  such as `te/device/main///cmd/software-update/+` or `te/device/main/service/+/cmd/restart/+`
+- the list of states
+- for each state:
+  - the state name as defined by the operation API
+  - the name of the state owner which is the process that is responsible for advancing a command at this state
+    (the default being `tedge`, meaning that thin-edge is responsible for running the state checks and actions)
+  - the set of states which can be an outcome for this state actions
+  - possible extra instructions on how to process the command at this stage, e.g.
+    - run a script
+    - emit an event
+    - raise an alarm
+  
+```toml title="file: workflow-example.toml"
+target = "te/device/main///cmd/configuration-update/+"
+
+[Init]
+  next = ["Download", "Failed"]
+  script = "/bin/schedule-configuration-update.sh"
+    
+[Download]
+  owner = tedge-configuration-plugin
+  next = ["Downloaded", "Failed"]
+
+[Downloaded]
+  script = "/bin/check-configuration-file.sh"
+  next = ["Install", "Failed"]
+
+[Install]
+  owner = tedge-configuration-plugin
+  next = ["Installed", "Failed"]
+        
+[Installed]
+  script = "/bin/check-configuration-update.sh"
+  next = ["Successful", "Failed"]
+
+[Successful]
+  event = "te/device/main///e/operations"
```

Comment:
It might be worthwhile considering supporting a way to refer to the target topic that the workflow instance was triggered on. This would support writing more generic workflows (if the user chooses to do this).

For example if the workflow is configured to react to any `configuration-update` command, however they would like for the event to be created on the target associated to the command, then the `event` field in the `[Success]` state would need some way of referring to the target topic.

```toml
target = "te/+/+/+/+/cmd/configuration-update"
// ...
[Success]
  event = "{identifier}/e/operations"
```